### PR TITLE
DBZ-3966 JsonTableChangeSerializer support serialization for defaultValue

### PR DESCRIPTION
**Pull request by Jiabao-Sun**
Both JsonTableChangeSerializer and ConnectTableChangeSerializer ignored defaultValue, hasDefaultValue, enumValues of ColumnImpl.

If we serialize and then deserialize the TableChanges, missing default value may cause:
```
org.apache.kafka.connect.errors.DataException: Invalid value: null used for required field
 ```

https://issues.redhat.com/browse/DBZ-3966

### CODE REVIEW

**Comment by Jiabao-Sun** [comment]
Failed tests need to be fixed.

**Comment by gunnarmorling** [comment]
@Jiabao-Sun, any insights on the test failures? Are you going to push a fix to this PR?

**Comment by Jiabao-Sun** [comment]
@gunnarmorling 
Sorry, I found that serializing `TableChange` is a bit more complicated than I thought.

Because we need to deal with some possible types of `ColumnImpl`'s default values ​​that are not recognized by `Value` such as `ByteBuffer` and `Short`.

Thanks for your reply and I'm going to push a fix to this PR.

**Comment by jpechane** [comment]
@Jiabao-Sun Cool, thanks! Do you think you might re-use `*DefaultValueConverter` classes to help with deserialization?

**Comment by Jiabao-Sun** [comment]
> @Jiabao-Sun Cool, thanks! Do you think you might re-use `*DefaultValueConverter` classes to help with deserialization?

@jpechane Thanks! I'll try with it.

**Comment by Jiabao-Sun** [comment]
@gunnarmorling @jpechane  Sorry for the late fix.

It's much difficult to serialize the `defaultValue` object to json directly, cause the data type of `defaultValue` may be unpredictable and non-serializable. In addition, forcibly serializing to json may also lose data types.

So, I added a new `defaultValueExpression` property of `ColumnImpl` to retain database original  default expression.
Serialize `defaultValueExpression` to json and use `DefaultValueConverter` to restore the defaultValue when deserialization.

It works but changes a lot. 
If there is a better plan, please suggest it.

**Comment by Jiabao-Sun** [comment]
@gunnarmorling @jpechane 

Failed tests were fixed and the PR is ready now.
Do you have any suggestions, or is there a better plan?

**Comment by jpechane** [comment]
@Jiabao-Sun I think this is nice approach, thanks!

@Naros What is the current support of default values in Oracle connector? Is it properly supported yet so this PR is vlaid for it or not?

**Comment by jpechane** [comment]
@Jiabao-Sun We are near the release of 1.7.0.Final. The PR will be merged right after the realse into 1.8 master.

**Comment by Jiabao-Sun** [comment]
@jpechane OK. Thanks a lot.

**Comment by Naros** [comment]
> What is the current support of default values in Oracle connector? Is it properly supported yet so this PR is vlaid for it or not?

@jpechane I vaguely remember that the shortcoming of the Oracle connector is we don't have a column default value listener and parser for the DDL.  So during snapshot we are able to get the values from the driver we don't capture and parse the value during streaming changes like we do for MySQL.

**Comment by gunnarmorling** [comment]
@Naros, can you log a separate issue for default value support for the Oracle connector then?

@jpechane, @Jiabao-Sun, other than Oracle support, is this PR good to go? If not, what's missing?

**Comment by jpechane** [comment]
@gunnarmorling This is good to go for me

**Comment by Naros** [comment]
> can you log a separate issue for default value support for the Oracle connector then?

@gunnarmorling Done, see [DBZ-4115](https://issues.redhat.com/browse/DBZ-4115).  I could have sworn we had one previously but I didn't see one currently opened.

**Comment by Jiabao-Sun** [comment]
Conflicts resolved.

**Comment by Jiabao-Sun** [comment]
Hi @jpechane, if you are available, please help review it.

File: debezium-core/src/main/java/io/debezium/relational/ColumnImpl.java [hunk 1]
```diff
@@ -133,8 +133,12 @@ public boolean isGenerated() {
...
 
     @Override
+    public Optional<String> defaultValueExpression() {
+        return Optional.ofNullable(defaultValueExpression);
+    }
+
     public Object defaultValue() {
-        return defaultValue;
+        return null;
     }
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/ColumnImpl.java:142]: Looks like a left-over?

Also `hasDefaultValue()` should not be needed any longer; as we now can differentiate whether a default value has been provided or not by examining the `Optional` returned from `defaultValueExpression()`. The explicit `hasDefaultValue()` method was needed before as we could not tell differentiate between no default value given or a default value of `null`.

**Review by gunnarmorling — COMMENTED** [review]
Thanks a lot for the rework, @Jiabao-Sun!

I think we're getting there. One more comment inline, looks like things can be simplified some more. I'll take another look tomorrow my time, but I don't see any more large road blocks.

File: debezium-core/src/main/java/io/debezium/relational/ColumnImpl.java (hunk 1 above)
Comment (Jiabao-Sun) [reply debezium-core/src/main/java/io/debezium/relational/ColumnImpl.java]: @gunnarmorling 

The left-over has been cleaned up.

I think we should keep `hasDefaultValue` cause there are scenarios for implicit default values.

Usually, `defaultValueExpression` is null and `hasDefaultValue` is false.
But in the implicit case, `defaultValueExpression` is null but `hasDefaultValue` should be true.

In a word, the null `defaultValueExpression` does not directly indicate whether the column has a default value.

https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html

Comment (Jiabao-Sun) [reply debezium-core/src/main/java/io/debezium/relational/ColumnImpl.java]: If this scenario needs to be concerned, then the `hasDefaultValue` should also be serialized in `TableChangeSerializer`.

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/ColumnImpl.java]: I see, that's a good point. Indeed we are exporting some "implicit default values" in certain cases, and my current thinking is we probably should not do this. But it seems largely independent from the matter at hand and fixing this also would be sort of a breaking change, so it's something we should leave for Debezium 2.0 (I'll add this to the list of things to be considered for that, see [DBZ-3899](https://issues.redhat.com/browse/DBZ-3899)). No re serializing `hasDefaultValue`, I'd prefer if we'd not do that, given we should get rid of this entire implicit default value business at all. Admittedly, there'll be an inconsistency when reading back the schema history for such value, but I feel that's an acceptable limitation.

Comment (Jiabao-Sun) [reply debezium-core/src/main/java/io/debezium/relational/ColumnImpl.java]: Thanks @gunnarmorling for the detailed explanation.
I will remove all the `hasDefaultValue`.

**Comment by gunnarmorling** [comment]
@Jiabao-Sun, pushed one more commit to return `Optional<String>` from `ColumnEditor`, too. If that passes CI, I think we can merge this PR. I'll create a follow-up draft PR, so to show how removing `hasDefaultValue` would look like. As said, that's something we likely want to keep for 2.0 though. Still going to share that preview PR, so to facilitate early feedback on it. In fact, one could also argue that the current behavior isn't correct actually, and a fix is warranted.

**Comment by Jiabao-Sun** [comment]
@gunnarmorling
By the way, if we still have to keep `hasDefaultValue` now, I recommend to serialize it in the `TableChangeSerializer` as well. Because ignoring it may cause inconsistent deserialization results eg. the column has implicit default value `null`.

File: debezium-core/src/main/java/io/debezium/relational/ColumnImpl.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/ColumnImpl.java]: Ok, I'm just not sure whether we can or should do this at this point. Here's a PR where I did that (just look at the last commit): https://github.com/debezium/debezium/pull/2887. It changes some assertions, i.e. behavior exposed to consumers is slightly modified in some cases. @jpechane expressed concerns about doing this in a release like 1.8 (as opposed to a major release like 2.0). @jpechane, can you take a look at that last commit and the assertions I've changed; as stated below, I think one could also argue that the previous behavior actually was incorrect when it comes to those implicit default values.

**Comment by gunnarmorling** [comment]
> By the way, if we still have to keep hasDefaultValue now, I recommend to serialize it in the TableChangeSerializer as well.

Yes, you are right. So I think we have two options at this point:

* Keep `hasDefaultValue` and also serialize it
* Do what I've sketched in that PR linked above, which would get rid of `hasDefaultValue` altogether

@Jiabao-Sun, @jpechane, which option would you prefer? I'm on the fence, I think I'm slighly leaning towards removing it, as I feel it's more about rectifying previous, buggy behavior rather than breaking correct, expected behavior. But I'm open for both options really.

**Comment by Jiabao-Sun** [comment]
@gunnarmorling 

Personally, I prefer the first option.
We can keep it in the minor version and try to remove it completely in the major version.

**Comment by gunnarmorling** [comment]
Ok, so let's do this then. Thanks a lot for iterating over this with us, really appreciating your efforts, @Jiabao-Sun! Are you going to push another commit for adding that serialization work? I'll log an issue describing the follow-up work, also linking to that branch of mine with the preliminary work for it. Thx!

**Comment by Jiabao-Sun** [comment]
OK, I will push another commit to complete it.

**Comment by gunnarmorling** [comment]
Rebased and applied. Thanks a lot, @Jiabao-Sun!

**Comment by gunnarmorling** [comment]
I've logged https://issues.redhat.com/browse/DBZ-4239 as a follow-up for removing `hasDefaultValue()` altogether.

**Comment by Jiabao-Sun** [comment]
Thanks, Cheers!

**Pull request closed without merging**